README.md
# FARM
**F**oreground **A**ll-scale **R**adio **M**odeller.
## Requirements
- python 3.6+ (f-string dependency)
### Python modules
- [astropy](https://docs.astropy.org/en/stable/) (developed with v5.0)
- [h5py](https://docs.h5py.org/en/stable/) (developed with v3.6.0)
- [jupyter](https://jupyter.org/) (developed with v1.0.0)
- [matplotlib](https://matplotlib.org/) (developed with v3.5.0)
- [numpy](https://numpy.org/) (developed with v1.22.1)
- [oskarpy](https://fdulwich.github.io/oskarpy-doc) (no version information available)
- [powerbox](https://powerbox.readthedocs.io/en/latest/) (developed with v0.6.1)
- [py21cmfast](https://21cmfast.readthedocs.io/en/latest/reference/py21cmfast.html) (no version information available)
- [pygdsm](https://github.com/telegraphic/pygdsm) (no version information available)
- [reproject](https://reproject.readthedocs.io/en/stable/) (developed with v0.8)
- [scipy](https://scipy.org/) (developed with v1.7.3)
- [toml](https://github.com/uiri/toml) (developed with v0.10.2)

### System
- [casa](https://casa.nrao.edu/) (developed with v6.4.0.16)
- [gsl](https://anaconda.org/conda-forge/gsl) (developed with v2.7)
- [miriad](https://www.atnf.csiro.au/computing/software/miriad/) (developed with v20220120)
- [oskar](https://ska-telescope.gitlab.io/sim/oskar/python/quickstart.html) (developed with v2.8.2)

## Installation help

For the python modules detailed above (and their dependencies (e.g. `gsl`), installation within a virtual environment is a possible method. Instructions for virtual environment creation and installing those dependencies listed above is shown below (requires a conda insallation).

**NOTE**: Installation of `py21cmFAST` on MacOSX leads to issues during compilation with the error, `clang: error: unsupported option '-fopenmp'` being thrown when installing via `pip install `. To properly resolve this and install `py21cmFAST` is detailed [here](https://github.com/21cmfast/21cmFAST/issues/84).

```commandline
conda create -n farm python=3
conda activate farm

# General package requirements
conda install astropy gsl h5py jupyter matplotlib numpy reproject scipy toml

# py21cmFAST
# OpenMP enabled libraries of fftw are available through conda-forge, using "conda install fftw" does not include openmp threads
conda install -c conda-forge fftw
conda install -c conda-forge/label/cf201901 gcc
conda install -c conda-forge 21cmFAST

# powerbox
conda install -c conda-forge pyfftw
conda install pip git
pip install git+git://github.com/steven-murray/powerbox.git

# Global diffuse sky model
pip install git+https://github.com/telegraphic/pygdsm

# OSKAR
pip install 'git+https://github.com/OxfordSKA/OSKAR.git@master#egg=oskarpy&subdirectory=python'
```

### A note on miriad and casa
Both miriad and casa limit the possible number of antennae in an array.

## Basic Use
### As a CLI application
```commandline
python3 /path/to/farm/main.py /path/to/config.toml
```
### As an imported library
```python
from pathlib import Path
import astropy.units as u
from astropy.coordinates import SkyCoord
import farm

dims  = 512, 512  # n_pixels in x/R.A., n_pixels in y/declination
field_of_view = 8.  # deg
cell_size = field_of_view / dims[0]  # deg

coord = SkyCoord("01:02:03.4", "05:06:07.89",
                 frame='fk5', unit=(u.hourangle, u.degree))

# Use MHD model for Galactic small-scale structure SkyComponent. This is loaded 
# from data/Gsynch_SKAs.fits directly and thus its frequency information is 
# fixed.
gssm = farm.SkyComponent.load_from_fits(fitsfile=farm.DATA_FILES['MHD'],
                                        name='GSSM', cdelt=cell_size,
                                        coord0=coord)

# Use GSM2016 model (Zheng et al, 2016) for Galactic diffuse-scale structure 
# SkyComponent. Add the same frequencies as the GSSM sky component to enable 
# combination.
gdsm = farm.SkyComponent(dims, cell_size, coord,
                         tb_func=farm.tb_functions.gdsm2016_t_b)
gdsm.add_frequency(gssm.frequencies)

# Create SkyModel instance which will handle the combination of the various 
# SkyComponent instances. Add frequency information to it
skymodel = farm.SkyModel(dims, cell_size, coord, gssm.frequencies)

# Add individual SkyComponent instances to the SkyModel
skymodel += (gssm, gdsm)  # Equivalent to skymodel.add_component((gssm, gdsm))

# Write the sky model to a .fits file
skymodel.write_fits(Path("test_skymodel_Inu.fits"), unit='JY/SR')
```


temp_input_image_ppW5jiFMne.im/history
CUNIT1  = 'deg     '
CUNIT2  = 'deg     '
CUNIT3  = 'Hz      '
FITS: Miriad fits: Revision 1.36, 2021/06/03 07:09:31 UTC
FITS: Executed on: 2022-03-30T15:01:22.0
FITS: Command line inputs follow:
FITS:   in=/var/folders/pq/0z61lr4x529bph79sd2mx3dhtw3wr6/T/temp_input_image_ppW5jiFMne_im_temp_WQ8qIAf3XA.fits
FITS:   out=temp_input_image_ppW5jiFMne.im
FITS:   op=xyin


testing.py
from pathlib import Path
import astropy.units as u
from astropy.coordinates import SkyCoord
import farm

dims  = 512, 512  # n_pixels in x/R.A., n_pixels in y/declination
field_of_view = 8.  # deg
cell_size = field_of_view / dims[0]  # deg

coord = SkyCoord("01:02:03.4", "05:06:07.89",
                 frame='fk5', unit=(u.hourangle, u.degree))

# Use MHD model for Galactic small-scale structure SkyComponent. This is loaded
# from data/Gsynch_SKAs.fits directly and thus its frequency information is
# fixed.
gssm = farm.SkyComponent.load_from_fits(fitsfile=farm.DATA_FILES['MHD'],
                                        name='GSSM', cdelt=cell_size,
                                        coord0=coord)

# Use GSM2016 model (Zheng et al, 2016) for Galactic diffuse-scale structure
# SkyComponent. Add the same frequencies as the GSSM sky component to enable
# combination.
gdsm = farm.SkyComponent(name='GDSM', npix=dims, cdelt=cell_size, coord0=coord,
                         tb_func=farm.tb_functions.gdsm2016_t_b)
gdsm.add_frequency(gssm.frequencies)

# Create SkyModel instance which will handle the combination of the various
# SkyComponent instances. Add frequency information to it
skymodel = farm.SkyModel(dims, cell_size, coord, gssm.frequencies)

# Add individual SkyComponent instances to the SkyModel
skymodel += (gssm, gdsm)  # Equivalent to skymodel.add_component((gssm, gdsm))

# Write the sky model to a .fits file
skymodel.write_fits(Path("test_skymodel_Inu.fits"), unit='JY/SR')
